### ISSUE TITLE
UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'catch' of undefined

### ISSUE THREAD

**Issue by VandeurenGlenn** [libp2p/js-libp2p-pubsub#58]
No clue what's going one here, unless I overlooked something evrything is catched also.

![image](https://user-images.githubusercontent.com/8685280/87256275-6579bb00-c491-11ea-86ed-5aaf4aceb818.png)

**Comment by VandeurenGlenn** [comment]
@jacobheun
https://github.com/libp2p/js-libp2p-pubsub/commit/f86dfbcc967c4cc5d01d7e04545ff0d0acc28d59#diff-0ea7f12adbcda08b8aecde80363ce603R112

Needs to be ?
```js await pipe```

**Comment by jacobheun** [comment]
@VandeurenGlenn are you able to reproduce this regularly or is it intermittent? 
What version of libp2p and Node.js are you using? Any additional context to help replicate this would be helpful.


The `await` isn't necessary as the returned value from pipe should be a Promise from calling `conn.sink()` at the end of the pipeline. My guess is that the new stream, https://github.com/libp2p/js-libp2p-pubsub/blob/e2ac4191d16200ad29f74aa4976cb9edffe0d675/src/index.js#L193, isn't returning a proper stream. If it's `sink` method is not asynchronous for some reason, undefined could be returned causing this.



_duplicates https://github.com/libp2p/js-libp2p/issues/694 in js-libp2p_

**Comment by VandeurenGlenn** [comment]
@jacobheun 
Node - v14.0.0
```js
"ipfs": "^0.47.0",
"libp2p": "^0.28.5",
"libp2p-delegated-content-routing": "^0.5.0",
"libp2p-delegated-peer-routing": "^0.5.0",
"libp2p-kad-dht": "^0.19.9",
```

The strange thing is, when I test my module it doesn't occur, but when installing as a dependency it does everytime.

**Comment by jacobheun** [comment]
> The strange thing is, when I test my module it doesn't occur, but when installing as a dependency it does everytime

This could be due to cacheing of your node_modules, such as your package-lock keeping an unaffected version. 

Not sure if it's related by I noticed some minor differences in node12 v node14 streams this weekend. I'll do some more testing in 14 to see if I can replicate more easily than 12.

**Comment by VandeurenGlenn** [comment]
Was thinking the same about the caching, trying to find where it broke, but cpu kinda weak.
I checked already on v12 and does the same.

For now no luck downgrading and reinstalling.

![image](https://user-images.githubusercontent.com/8685280/87306415-47a36900-c518-11ea-8ad7-66c87f3ae69c.png)

**Comment by jacobheun** [comment]
Is this just using IPFS, or Is the code you're using public anywhere? If you're getting this regularly it would be helpful for us reproducing it locally.

**Comment by VandeurenGlenn** [comment]
```sh
npm i ipfs@0.46.0 libp2p-pubsub@0.4.7
```
```js
const Ipfs = require('ipfs');

(async () => {
  const ipfs = await Ipfs.create()
})()
```

Should be able to reproduce everytime

**Comment by jacobheun** [comment]
Thanks that helped, this is a bug with how metrics wraps streams. I'll create a fix for libp2p 0.27 and 0.28. I'll close this and track at https://github.com/libp2p/js-libp2p/issues/694 since it's a libp2p core issue.
